tracer/src/Datadog.Trace/Ci/GitCommandHelper.cs

```diff
@@ -188,6 +189,222 @@ static List<FileCoverageInfo> ParseGitDiff(string diffOutput)
+            }
+
+            // 5. Compute metrics for each branch
+            var metrics = new List<Tuple<string, string, int, int>>(); // Branch, MergeBase, Behind, Ahead
+
+            foreach (var branch in branches)
+            {
+                // Find merge-base (common ancestor)
+                var mergeBaseOutput = RunGitCommand(
+                    workingDirectory,
+                    $"merge-base {branch} {targetBranch}",
+                    MetricTags.CIVisibilityCommands.MergeBase);
+
+                if (mergeBaseOutput is not { ExitCode: 0 } || string.IsNullOrWhiteSpace(mergeBaseOutput?.Output))
+                {
+                    continue; // Skip if no common history
+                }
+
+                var mergeBaseSha = mergeBaseOutput!.Output.Trim();
+
+                // Count commits ahead and behind
+                var revListOutput = RunGitCommand(
+                    workingDirectory,
+                    $"rev-list --left-right --count {branch}...{targetBranch}",
+                    MetricTags.CIVisibilityCommands.RevList);
+
+                if (revListOutput is not { ExitCode: 0 } || string.IsNullOrWhiteSpace(revListOutput?.Output))
+                {
+                    continue; // Skip if it cannot get commit counts
+                }
+
+                var counts = revListOutput!.Output.Split(new[] { '\t', ' ' }, StringSplitOptions.RemoveEmptyEntries);
+                if (counts.Length != 2 ||
+                    !int.TryParse(counts[0], out var behind) ||
+                    !int.TryParse(counts[1], out var ahead))
+                {
+                    continue; // Skip if unexpected format or cannot parse counts
+                }
+
+                metrics.Add(Tuple.Create(branch, mergeBaseSha, behind, ahead));
+            }
+
+            if (metrics.Count == 0)
+            {
+                Log.Warning("GitCommandHelper: No metrics could be computed for any candidate branch");
+                return null;
+            }
+
+            // 6. Sort by the "behind" metric (ascending) to find the best base branch
+            metrics.Sort((a, b) => a.Item3.CompareTo(b.Item3)); // Sort by Behind (Item3)
+
+            // Find candidates with the minimum "behind" value
+            int bestBehind = metrics[0].Item3;
+            var bestCandidates = metrics.Where(m => m.Item3 == bestBehind).ToList();
+
+            // If multiple candidates with same "behind" value, prioritize default branch
+            var bestCandidate = bestCandidates[0];
+
+            if (bestCandidates.Count > 1 && !string.IsNullOrEmpty(defaultBranch))
+            {
+                foreach (var candidate in bestCandidates)
+                {
+                    if (candidate.Item1 == defaultBranch || candidate.Item1 == $"{remoteName}/{defaultBranch}")
+                    {
+                        bestCandidate = candidate;
+                        break;
+                    }
+                }
+            }
```

Comment:
This can be rewritten to reduce memory allocations. For example you could use `.Any` before trying to do the `.Where().List()` that allocates a new empty list if there's no behind collision.